dist=loadscope scheduling hangs on test class with only one test

With pytest 3.2.3, when I tried using --dist=loadscope with two ssh nodes, the test framework simply hung without even trying to run any tests.  The same environment works fine with the other schedulers.

Part of the traceback (seems exactly the same as #230 ):
        while not self.session_finished:
>           self.loop_once()

The hang seems to occur when running a test class that has only one test function.
I eventually tracked down a simple change that seems to fix it: on the following line, expect only one item, not two:

https://github.com/pytest-dev/pytest-xdist/blob/master/xdist/remote.py#L62

I can't tell what the change might break, though.

If I send SIGINT, the parent reports this:
```
run.venv/lib/python2.7/site-packages/pytest-3.2.3-py2.7.egg/_pytest/main.py:110: in wrap_session
    session.exitstatus = doit(config, session) or 0
run.venv/lib/python2.7/site-packages/pytest-3.2.3-py2.7.egg/_pytest/main.py:146: in _main
    config.hook.pytest_runtestloop(session=session)
run.venv/lib/python2.7/site-packages/pytest_xdist-1.20.1-py2.7.egg/xdist/dsession.py:114: in pytest_runtestloop
    self.loop_once()
run.venv/lib/python2.7/site-packages/pytest_xdist-1.20.1-py2.7.egg/xdist/dsession.py:128: in loop_once
    eventcall = self.queue.get(timeout=2.0)
[REDACTED]/lib/python2.7/Queue.py:177: in get
    self.not_empty.wait(remaining)
[REDACTED]/lib/python2.7/threading.py:359: in wait
    _sleep(delay)
E   KeyboardInterrupt
```

The slave / executor reports this:
```
run.venv/lib/python2.7/site-packages/pytest-3.2.3-py2.7.egg/_pytest/main.py:110: in wrap_session
    session.exitstatus = doit(config, session) or 0
run.venv/lib/python2.7/site-packages/pytest-3.2.3-py2.7.egg/_pytest/main.py:146: in _main
    config.hook.pytest_runtestloop(session=session)
<remote exec>:52: in pytest_runtestloop
    ???
<string>:732: in receive
    ???
[REDACTED]/lib/python2.7/Queue.py:168: in get
    self.not_empty.wait()
[REDACTED]/lib/python2.7/threading.py:340: in wait
    waiter.acquire()
E   KeyboardInterrupt
```

Note that [REDACTED] denotes where the python's source lives, outside the virtualenv.

--- a/changelog/257.bugfix
+++ b/changelog/257.bugfix
@@ -1,0 +1,1 @@
+Fix issue when using ``loadscope`` or ``loadfile`` where tests would fail to start if the first scope had only one test.

--- a/testing/acceptance_test.py
+++ b/testing/acceptance_test.py
@@ -888,6 +888,38 @@ class TestLoadScope:
         assert get_workers_and_test_count_by_prefix(
             'test_a.py::TestB', result.outlines) in ({'gw0': 10}, {'gw1': 10})
 
+    def test_module_single_start(self, testdir):
+        test_file1 = """
+            import pytest
+            def test():
+                pass
+        """
+        test_file2 = """
+            import pytest
+            def test_1():
+                pass
+            def test_2():
+                pass
+        """
+        testdir.makepyfile(
+            test_a=test_file1,
+            test_b=test_file1,
+            test_c=test_file2
+        )
+        result = testdir.runpytest('-n2', '--dist=loadscope', '-v')
+        a = get_workers_and_test_count_by_prefix('test_a.py::test',
+                                                 result.outlines)
+        b = get_workers_and_test_count_by_prefix('test_b.py::test',
+                                                 result.outlines)
+        c1 = get_workers_and_test_count_by_prefix('test_c.py::test_1',
+                                                  result.outlines)
+        c2 = get_workers_and_test_count_by_prefix('test_c.py::test_2',
+                                                  result.outlines)
+        assert a in ({'gw0': 1}, {'gw1': 1})
+        assert b in ({'gw0': 1}, {'gw1': 1})
+        assert a.items() != b.items()
+        assert c1 == c2
+
 
 class TestFileScope:
 
@@ -942,6 +974,38 @@ class TestFileScope:
             test_a_workers_and_test_count in ({'gw0': 0}, {'gw1': 10})
         assert test_b_workers_and_test_count in ({'gw0': 10}, {'gw1': 0}) or \
             test_b_workers_and_test_count in ({'gw0': 0}, {'gw1': 10})
+
+    def test_module_single_start(self, testdir):
+        test_file1 = """
+            import pytest
+            def test():
+                pass
+        """
+        test_file2 = """
+            import pytest
+            def test_1():
+                pass
+            def test_2():
+                pass
+        """
+        testdir.makepyfile(
+            test_a=test_file1,
+            test_b=test_file1,
+            test_c=test_file2
+        )
+        result = testdir.runpytest('-n2', '--dist=loadfile', '-v')
+        a = get_workers_and_test_count_by_prefix('test_a.py::test',
+                                                 result.outlines)
+        b = get_workers_and_test_count_by_prefix('test_b.py::test',
+                                                 result.outlines)
+        c1 = get_workers_and_test_count_by_prefix('test_c.py::test_1',
+                                                  result.outlines)
+        c2 = get_workers_and_test_count_by_prefix('test_c.py::test_2',
+                                                  result.outlines)
+        assert a in ({'gw0': 1}, {'gw1': 1})
+        assert b in ({'gw0': 1}, {'gw1': 1})
+        assert a.items() != b.items()
+        assert c1 == c2
 
 
 def parse_tests_and_workers_from_output(lines):

--- a/testing/acceptance_test.py
+++ b/testing/acceptance_test.py
@@ -889,6 +889,7 @@ class TestLoadScope:
             'test_a.py::TestB', result.outlines) in ({'gw0': 10}, {'gw1': 10})
 
     def test_module_single_start(self, testdir):
+        """Fix test suite never finishing in case all workers start with a single test (#277)."""
         test_file1 = """
             import pytest
             def test():
@@ -976,6 +977,7 @@ class TestFileScope:
             test_b_workers_and_test_count in ({'gw0': 0}, {'gw1': 10})
 
     def test_module_single_start(self, testdir):
+        """Fix test suite never finishing in case all workers start with a single test (#277)."""
         test_file1 = """
             import pytest
             def test():

--- a/xdist/scheduler/loadscope.py
+++ b/xdist/scheduler/loadscope.py
@@ -380,6 +380,10 @@ class LoadScopeScheduling:
         for node in self.nodes:
             self._assign_work_unit(node)
 
+        # Ensure nodes start with at least two work units if possible
+        for node in self.nodes:
+            self._reschedule(node)
+
         # Initial distribution sent all tests, start node shutdown
         if not self.workqueue:
             for node in self.nodes:

--- a/xdist/scheduler/loadscope.py
+++ b/xdist/scheduler/loadscope.py
@@ -380,7 +380,7 @@ class LoadScopeScheduling:
         for node in self.nodes:
             self._assign_work_unit(node)
 
-        # Ensure nodes start with at least two work units if possible
+        # Ensure nodes start with at least two work units if possible (#277)
         for node in self.nodes:
             self._reschedule(node)
 

